SDL/Event.cu

```diff
@@ -640,57 +647,82 @@ public:
     }
 };
 
-
-__global__ void hitLoopKernel(
-                            uint16_t Endcap, // Integer corresponding to endcap in module subdets
-                            uint16_t TwoS, // Integer corresponding to TwoS in moduleType
-                            int nHits, // Total number of hits in event
-                            unsigned int nModules, // Number of modules
-                            unsigned int nEndCapMap, // Number of elements in endcap map
-                            unsigned int* geoMapDetId, // DetId's from endcap map
-                            float* geoMapPhi, // Phi values from endcap map
-                            struct SDL::modules *modulesInGPU,
-                            struct SDL::hits *hitsInGPU)
+class hitLoopKernel
 {
-    for( int ihit = blockIdx.x * blockDim.x + threadIdx.x; ihit < nHits; ihit += blockDim.x * gridDim.x )
+public:
+    ALPAKA_NO_HOST_ACC_WARNING
+    template<
+        typename TAcc,
+        typename TIdx>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const & acc,
+        uint16_t Endcap, // Integer corresponding to endcap in module subdets
+        uint16_t TwoS, // Integer corresponding to TwoS in moduleType
+        unsigned int nModules, // Number of modules
+        unsigned int nEndCapMap, // Number of elements in endcap map
+        unsigned int* geoMapDetId, // DetId's from endcap map
+        float* geoMapPhi, // Phi values from endcap map
+        struct SDL::modules *modulesInGPU,
+        struct SDL::hits *hitsInGPU,
+        TIdx const & nHits) const // Total number of hits in event
+    -> void
     {
-        float ihit_x = hitsInGPU->xs[ihit];
-        float ihit_y = hitsInGPU->ys[ihit];
-        float ihit_z = hitsInGPU->zs[ihit];
-        int iDetId = hitsInGPU->detid[ihit];
 
-        hitsInGPU->rts[ihit] = sqrt(ihit_x*ihit_x + ihit_y*ihit_y);
-        hitsInGPU->phis[ihit] = SDL::phi(ihit_x,ihit_y,ihit_z);
-        hitsInGPU->etas[ihit] = ((ihit_z>0)-(ihit_z<0))*acosh(sqrt(ihit_x*ihit_x+ihit_y*ihit_y+ihit_z*ihit_z)/hitsInGPU->rts[ihit]);
+        TIdx const gridThreadIdx(alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]);
+        TIdx const threadElemExtent(alpaka::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
+        TIdx const threadFirstElemIdx(gridThreadIdx * threadElemExtent);
 
-        int found_index = binary_search(modulesInGPU->mapdetId, iDetId, nModules);
-        uint16_t lastModuleIndex = modulesInGPU->mapIdx[found_index];
+        if(threadFirstElemIdx < nHits)
+        {
+            TIdx const threadLastElemIdx(threadFirstElemIdx+threadElemExtent);
+            TIdx const threadLastElemIdxClipped((nHits > threadLastElemIdx) ? threadLastElemIdx : nHits);
 
-        hitsInGPU->moduleIndices[ihit] = lastModuleIndex;
+            for(TIdx ihit(threadFirstElemIdx); ihit<threadLastElemIdxClipped; ++ihit)
```

Comment:
As per this comment [here](https://github.com/SegmentLinking/TrackLooper/pull/209#discussion_r996465189), is there any practical reason to complicate the code like this, instead of writing cuda-like loops? Especially because on the GPUs we don't work on the element-level objects.

Reply:
Isn't the hope that we'll be able to eventually run this algo on CPU's as well? It would be a pain to have to generalize the code in the future if we find that we need to take advantage of elements. Although it's unclear how useful elements are. See this issue for example: https://github.com/alpaka-group/alpaka/issues/1824

Reply:
Isn't the hope that we'll be able to eventually run this algo on CPU's as well? It would be a pain to have to generalize the code in the future if we find that we need to take advantage of elements. Although it's unclear how useful elements are.